prov/verbs/src/verbs_ep.c

```diff
@@ -966,15 +1034,17 @@ static int vrb_ep_enable(struct fid_ep *ep_fid)
 			return -FI_EINVAL;
 		}
 
-		ret = rdma_create_qp(ep->id, domain->pd, &attr);
-		if (ret) {
-			VRB_WARN_ERRNO(FI_LOG_EP_CTRL, "rdma_create_qp");
-			return -errno;
-		}
+		if (ep->state == VRB_REQ_RCVD) {
```

Comment:
`vrb_eq_addr_resolved_event()` only creates the QP for the connection initiator. The target side needs to create the QP when the EP is enabled, after `RDMA_CM_EVENT_CONNECT_REQUEST` is received (EP state is VRB_REQ_RCVD in this case). 
To be honest, I don't like this check neither. Do you have a better suggestion? Is there a way to determine what side enables the EP, for example based on the content of ep->id?

Reply:
I changed the check and the code now uses `ep->id->verbs` to figure out what side (`server` or `client`) executes the function. I looked at rdma-core and it seems that `ep->id->verbs` is set only when the addr or the route are resolved:
https://github.com/linux-rdma/rdma-core/blob/0cf342c301dfef1ba31c7ab763682ed0c59e7763/librdmacm/cma.c#L618

Reply:
id->verbs field is ibv_context and its creation/assignment is not related to addr/route resolution.

Please consider creating QP in RDMA_CM_EVENT_CONNECT_REQUEST handler see https://github.com/linux-rdma/rdma-core/blob/master/librdmacm/man/rdma_cm.7

Reply:
We cannot move QP creation in `RDMA_CM_EVENT_CONNECT_REQUEST` handler because `rdma_create_qp()` requires to set the QP init attributes (`ibv_qp_init_attr`), and these attributes are known only when the EP is enabled. The EP doesn't even exist when the event `RDMA_CM_EVENT_CONNECT_REQUEST` is received.

> id->verbs field is ibv_context and its creation/assignment is not related to addr/route resolution.

I agree with your statement, but in practice and according to the code, id->verbs set when the addr/route is resolved. 

An alternative solution to find out if the route was resolved would be to check if `rdma_cm_id::route::num_paths` is > 0 or `rdma_cm_id::route::path_rec` is != NULL.

Reply:
I'm still running tests, trying out large MPI jobs.  It is taking a bit of time.